File: Week_06/cpp/src/_204_CountPrimes.cpp
#include <vector>
using namespace std;
class Navie
{
public:
    int countPrimes(int n)
    {
        int ret = 0;
        auto isPrime = [](int x) -> bool {
            for (int i = 2; i * i <= x; i++)
            {
                if (x % i == 0)
                    return false;
            }
            return true;
        };
        for (int i = 2; i < n; i++)
        {
            if (isPrime(i))
                ret++;
        }
        return ret;
    }
};
class UsingEratostheness
{
public:
    // O(nloglogn) O(n)
    // 枚举没有考虑数与数之间的关联性
    // 如果x是质数，那么大于x的倍数 2x,3x...都是合数
    int countPrimes(int n)
    {
        vector<int> isPrime(n, 1);
        int ret = 0;
        for (int i = 2; i < n; i++)
        {
            if (isPrime[i])
            {
                ret++;
                for (long j = i; j * i < n; j++)
                {
                    isPrime[j * i] = 0;
                }
            }
        }
        return ret;
    }
};

File: Week_06/kotlin/src/_204_CountPrimes.kt
class _204_CountPrimes_ {
    class UsingEratostheness {
        fun countPrimes(n: Int): Int {
            val isPrime = BooleanArray(n) { true }
            var ret = 0
            for (i in 2 until n) {
                if (isPrime[i]) {
                    ret++
                    var j = i
                    while (j * i.toLong() < n) {
                        isPrime[j * i] = false
                        j++
                    }
                }
            }
            return ret
        }
    }

    class Naive {
        fun countPrimes(n: Int): Int {
            var ret = 0
            fun isPrime(x: Int): Boolean {
                var y = 2L
                while (y * y <= x) {
                    if (x % y == 0L) return false
                    y++
                }
                return true
            }
            for (i in 2 until n) {
                if (isPrime(i)) ret++
            }
            return ret
        }
    }
}